packages/firestore/test/unit/local/query_engine.test.ts

```diff
@@ -281,498 +290,713 @@ function genericQueryEngineTest(
+      await indexManager.updateCollectionGroup(
+        'coll',
+        newIndexOffsetFromDocument(doc2)
+      );
+
+      await addDocument(doc3);
+      await addMutation(setMutation('coll/d', { 'foo': true }));
+
+      const queryWithFilter = queryWithAddedFilter(
+        query('coll'),
+        filter('foo', '==', true)
+      );
+      const results = await expectOptimizedCollectionQuery(() =>
+        runQuery(queryWithFilter, SnapshotVersion.min())
+      );
+
+      verifyResult(results, [doc1, doc2, doc3, doc4]);
+    });
+
+    it('uses partial index for limit queries', async () => {
+      debugAssert(configureCsi, 'Test requires durable persistence');
+
+      const doc1 = doc('coll/1', 1, { 'a': 1, 'b': 0 });
+      const doc2 = doc('coll/2', 1, { 'a': 1, 'b': 1 });
+      const doc3 = doc('coll/3', 1, { 'a': 1, 'b': 2 });
+      const doc4 = doc('coll/4', 1, { 'a': 1, 'b': 3 });
+      const doc5 = doc('coll/5', 1, { 'a': 2, 'b': 3 });
+      await addDocument(doc1, doc2, doc3, doc4, doc5);
+
+      await indexManager.addFieldIndex(
+        fieldIndex('coll', { fields: [['a', IndexKind.ASCENDING]] })
+      );
+      await indexManager.updateIndexEntries(
+        documentMap(doc1, doc2, doc3, doc4, doc5)
+      );
+      await indexManager.updateCollectionGroup(
+        'coll',
+        newIndexOffsetFromDocument(doc5)
+      );
+
+      const q = queryWithLimit(
+        queryWithAddedFilter(
+          queryWithAddedFilter(query('coll'), filter('a', '==', 1)),
+          filter('b', '==', 1)
+        ),
+        3,
+        LimitType.First
+      );
+      const results = await expectOptimizedCollectionQuery(() =>
+        runQuery(q, SnapshotVersion.min())
+      );
+
+      verifyResult(results, [doc2]);
+    });
+
+    it('re-fills indexed limit queries', async () => {
+      debugAssert(configureCsi, 'Test requires durable persistence');
+
+      const doc1 = doc('coll/1', 1, { 'a': 1 });
+      const doc2 = doc('coll/2', 1, { 'a': 2 });
+      const doc3 = doc('coll/3', 1, { 'a': 3 });
+      const doc4 = doc('coll/4', 1, { 'a': 4 });
+      await addDocument(doc1, doc2, doc3, doc4);
+
+      await indexManager.addFieldIndex(
+        fieldIndex('coll', { fields: [['a', IndexKind.ASCENDING]] })
+      );
+      await indexManager.updateIndexEntries(
+        documentMap(doc1, doc2, doc3, doc4)
+      );
+      await indexManager.updateCollectionGroup(
+        'coll',
+        newIndexOffsetFromDocument(doc4)
+      );
+
+      await addMutation(patchMutation('coll/3', { 'a': 5 }));
+
+      const q = queryWithLimit(
+        queryWithAddedOrderBy(query('coll'), orderBy('a')),
+        3,
+        LimitType.First
+      );
+      const results = await expectOptimizedCollectionQuery(() =>
+        runQuery(q, SnapshotVersion.min())
+      );
+
+      verifyResult(results, [doc1, doc2, doc4]);
+    });
+
+    it('or query with in and not-in using index', async () => {
```

Comment:
`or query with in and not-in` test and `or query with in and not-in using index` test can be merged into one test (using `if (configureCsi)`.

Reply:
Good catch. This is updated.